openstack_controller/tests/test_unit_nova.py

```diff
@@ -1857,6 +1858,168 @@ def test_servers_metrics(aggregator, check, dd_run_check, metrics):
+        aggregator.assert_metric(
+            metric['name'],
+            count=metric.get('count'),
+            value=metric.get('value'),
+            tags=metric.get('tags'),
+            hostname=metric.get('hostname'),
+        )
+
+    if api_type == ApiType.REST:
+        args_list = []
+        for call in mock_http_get.call_args_list:
+            args, kwargs = call
+            params = kwargs.get('params', {})
+            project_id = params.get('project_id')
+            limit = params.get('limit')
+            args_list += [(args[0], project_id, limit)]
+        servers_url = 'http://127.0.0.1:8774/compute/v2.1/servers/detail'
+        assert (
+            args_list.count((servers_url, '1e6e233e637d4d55a50a62b63398ad15', paginated_limit))
+            == expected_api_calls_proj1
+        )
+        assert (
+            args_list.count((servers_url, '6e39099cccde4f809b003d9e0dd09304', paginated_limit))
+            == expected_api_calls_proj2
+        )
+
+    if api_type == ApiType.SDK:
+        assert (
+            connection_compute.servers.call_args_list.count(
+                mock.call(project_id='6e39099cccde4f809b003d9e0dd09304', details=True, limit=paginated_limit)
+            )
+            == expected_api_calls_proj1
+        )
+        assert (
+            connection_compute.servers.call_args_list.count(
+                mock.call(project_id='1e6e233e637d4d55a50a62b63398ad15', details=True, limit=paginated_limit)
+            )
+            == expected_api_calls_proj2
+        )
+
+
+@pytest.mark.parametrize(
+    ('mock_http_get', 'paginated_limit'),
+    [
+        pytest.param(
+            {
+                'mock_data': {
+                    '/compute/v2.1/servers/detail': {"servers": []},
+                }
+            },
+            1,
+            id='api empty servers',
+        ),
+        pytest.param(
+            {
+                'mock_data': {
+                    '/compute/v2.1/servers/detail': {"servers": [{"test": "attr"}], "servers_links": [{"rel": "next"}]},
+                }
+            },
+            1,
+            id='api one invalid server',
+        ),
+        pytest.param(
+            {
+                'mock_data': {
+                    '/compute/v2.1/servers/detail': {
+                        "servers": [{"test": "attr", "id": "test"}],
+                        "servers_links": [{"rel": "test"}],
+                    },
+                }
+            },
+            1,
+            id='api one invalid server link',
+        ),
+    ],
+    indirect=['mock_http_get'],
+)
+@pytest.mark.usefixtures('mock_http_get', 'mock_http_post', 'openstack_connection')
+def test_pagination_invalid_no_exception(openstack_controller_check, dd_run_check, paginated_limit):
+    paginated_instance = copy.deepcopy(configs.REST)
+    paginated_instance['paginated_limit'] = paginated_limit
+    check = openstack_controller_check(paginated_instance)
+    dd_run_check(check)
+
```

Comment:
I miss the assert block here. Should we assert no server metrics are reported?

Reply:
Good idea!